File: python/mirage/workspace/workspace/policy.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Callable
from typing import Any

from mirage.runtime.policy import (PolicyContext, PolicyDecision, PolicyError,
                                   PolicyFn, decide_line, parsed_commands)
from mirage.runtime.table import catch_all, runtime_bindings_for
from mirage.workspace.mount import MountRegistry
from mirage.workspace.session import Session
from mirage.workspace.workspace.runtimes import Runtimes

PrefixFn = Callable[[], list[str]]


class PolicyRouter:
    """Decides which runtime a typed line routes to.

    The order is: an inherited decision, then the ``execute()`` runtime
    argument, then the configured policy and any entry scripts. It
    reads the runtime entries and the registry's static bindings but
    owns no mutable workspace state, so the volatile parts (the policy
    callable, the current agent) arrive per call and a new step is
    added here rather than in the workspace.

    Args:
        registry (MountRegistry): carries the resolved static bindings.
        runtimes (Runtimes): the ordered runtime entries.
        mount_prefixes (PrefixFn): mount prefixes for the policy context.
    """

    def __init__(self, registry: MountRegistry, runtimes: Runtimes,
                 mount_prefixes: PrefixFn) -> None:
        self._registry = registry
        self._runtimes = runtimes
        self._mount_prefixes = mount_prefixes

    async def decide(
        self,
        ast: Any,
        command: str,
        runtime: str | None,
        provision: bool,
        session: Session,
        session_id: str,
        agent_id: str,
        policy: PolicyFn | None,
        inherited: PolicyDecision | None,
    ) -> PolicyDecision | None:
        """Resolve the routing decision for one typed line.

        Returns None when nothing decides (no runtime argument, no
        policy configured) so dispatch falls to the static bindings. A
        nested eval passes its typed line's decision as ``inherited``
        and keeps it: nested lines never re-route. Provision never
        routes.

        Args:
            ast: the parsed tree-sitter root node.
            command (str): the raw command line.
            runtime (str | None): the execute() runtime argument, which
                wins over the policy.
            provision (bool): whether this is a provision run.
            session (Session): the effective session (cwd, env).
            session_id (str): session hosting the line.
            agent_id (str): agent the line runs as.
            policy (PolicyFn | None): the workspace policy, if any.
            inherited (PolicyDecision | None): the calling line's
                decision, for nested evals.

        Raises:
            PolicyError: an unknown runtime name or a failing policy.
        """
        if inherited is not None:
            return inherited
        entries = self._runtimes.entries
        if runtime is not None:
            try:
                overlay = runtime_bindings_for(entries, runtime)
            except ValueError as exc:
                raise PolicyError(str(exc)) from exc
            return PolicyDecision(bindings={
                **self._registry.runtime_bindings,
                **overlay
            },
                                  fallback=catch_all(entries))
        if provision:
            return None
        has_scripts = any(entry.script is not None for entry in entries)
        if policy is None and not has_scripts:
            return None
        commands = parsed_commands(ast, self._registry.clis.names())
        ctx = PolicyContext(
            line=command,
            commands=commands,
            command=commands[0].command if commands else "",
            builtin=commands[0].builtin if commands else False,
            cwd=session.cwd,
            env=dict(session.env),
            session_id=session_id,
            agent_id=agent_id,
            mounts=tuple(self._mount_prefixes()),
        )
        try:
            return await decide_line(entries, policy, ctx,
                                     self._registry.runtime_bindings)
        except PolicyError:
            raise
        except (ValueError, ImportError) as exc:
            raise PolicyError(str(exc)) from exc

File: typescript/packages/core/src/workspace/workspace/policy.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  parsedCommands,
  decideLine,
  PolicyError,
  type PolicyContext,
  type PolicyDecision,
  type PolicyFn,
} from '../executor/policy/index.ts'
import { catchAll, runtimeBindingsFor, type Runtime } from '../executor/runtime.ts'
import type { TSNodeLike } from '../expand/variable.ts'
import type { MountRegistry } from '../mount/registry.ts'
import type { SessionManager } from '../session/manager.ts'
import type { ExecuteOptions } from './types.ts'
import type { Runtimes } from './runtimes.ts'

/**
 * The policy ladder for one typed line: runtime argument, policy,
 * scripts. Mirrors the Python `PolicyRouter` in `workspace/policy.py`.
 *
 * `decide` returns null when nothing decides (no runtime argument, no
 * policy configured) so dispatch falls to the static bindings; a nested
 * eval inherits the typed line's decision and never re-routes.
 */
export class PolicyRouter {
  private readonly registry: MountRegistry
  private readonly runtimes: Runtimes
  private readonly policy: PolicyFn | null
  private readonly sessions: SessionManager
  private readonly agentId: string | null
  private readonly visibleMounts: () => string[]

  constructor(
    registry: MountRegistry,
    runtimes: Runtimes,
    policy: PolicyFn | null,
    sessions: SessionManager,
    agentId: string | null,
    visibleMounts: () => string[],
  ) {
    this.registry = registry
    this.runtimes = runtimes
    this.policy = policy
    this.sessions = sessions
    this.agentId = agentId
    this.visibleMounts = visibleMounts
  }

  async decide(
    root: TSNodeLike,
    command: string,
    options: ExecuteOptions,
  ): Promise<PolicyDecision | null> {
    if (options.routingDecision !== undefined) return options.routingDecision
    if (options.runtime !== undefined) {
      let overlay: Record<string, Runtime>
      try {
        overlay = runtimeBindingsFor(this.runtimes.entries, options.runtime)
      } catch (caught) {
        throw new PolicyError(caught instanceof Error ? caught.message : String(caught), {
          cause: caught,
        })
      }
      return {
        bindings: Object.assign(
          Object.create(null) as Record<string, Runtime>,
          this.runtimes.bindings,
          overlay,
        ),
        fallback: catchAll(this.runtimes.entries),
      }
    }
    const hasScripts = this.runtimes.entries.some((entry) => entry.script !== undefined)
    if (this.policy === null && !hasScripts) return null
    const commands = parsedCommands(root, this.registry.clis.names())
    const sessionId = options.sessionId ?? this.sessions.defaultId
    const session = this.sessions.get(sessionId)
    const ctx: PolicyContext = {
      line: command,
      commands,
      command: commands[0]?.command ?? '',
      builtin: commands[0]?.builtin ?? false,
      cwd: options.cwd ?? session.cwd,
      env: { ...session.env, ...(options.env ?? {}) },
      sessionId,
      agentId: options.agentId ?? this.agentId ?? '',
      mounts: this.visibleMounts(),
    }
    return decideLine(this.runtimes.entries, this.policy, ctx, this.runtimes.bindings)
  }
}
